TypeScript Adoption Plan: get it started :tada:

chriskrycho opened the pull request:
[Rendered draft](https://github.com/chriskrycho/ember-rfcs/blob/ts-adoption/text/0000-ts-adoption-plan.md) – now open for comments and input from the TypeScript strike team for polishing before opening against Ember!

Looking for feedback especially on:

- the decision to have Ember Data type registries handled in a separate RFC
- the unresolved questions
- whether there are drawbacks I missed

chriskrycho committed:
TypeScript Adoption Plan: get it started :tada:

chriskrycho committed:
TS Adoption: add basic motivation

chriskrycho committed:
Start on 'How we teach this'

chriskrycho committed:
Add 'Drawbacks' (quoting #0724)

chriskrycho committed:
Start on 'Alternatives' section

chriskrycho committed:
Expand on 'Alternatives'

chriskrycho committed:
Remove GitHub-stripped `abbr` tags

chriskrycho committed:
Update outline (no numbers!)

chriskrycho committed:
Expand outline, add some TODOs

chriskrycho committed:
Document approach to classic features

chriskrycho committed:
Update outline

chriskrycho committed:
Nit: md doesn't work in inline HTML

chriskrycho committed:
Cover basics of SemVer

chriskrycho committed:
More details on how we teach SemVer

chriskrycho committed:
Add discussion of handling Classic types

chriskrycho committed:
Extend discussion/example for SemVer

chriskrycho committed:
Fix some typos

chriskrycho committed:
Another typo

chriskrycho committed:
Add (empty) strictness section, summarize SemVer section

chriskrycho committed:
Clarify the no new red squiggles rule

chriskrycho committed:
Clarify wording in no red squiggles rule

chriskrycho committed:
Finish an incomplete sentence.

chriskrycho committed:
Start strictness discussion

chriskrycho committed:
Fix example table

chriskrycho committed:
Further improve/clarify example table

chriskrycho committed:
More on stability, SemVer, and esp. LTS releases

chriskrycho committed:
Fix a typo

chriskrycho committed:
Fix a typo

chriskrycho committed:
Spec compliance

chriskrycho committed:
Add section: CLI Integration

chriskrycho committed:
Outline the detailed design

chriskrycho committed:
Start on MIgration from DT section

chriskrycho committed:
Add discussion of 'Edition support policy'

chriskrycho committed:
Discuss type registries

chriskrycho committed:
Add discussion of mechanics and docs for migration from DT

chriskrycho committed:
Slightly elaborate motivation

chriskrycho committed:
Add alternative on Classic Features support

chriskrycho committed:
Add basic section on CLI integration

chriskrycho committed:
Add another unresolved question

chriskrycho committed:
SemVer RFC tweak

chriskrycho committed:
Another unresolved question: re: ember-ultra-strict

chriskrycho committed:
Finish discussion of Strictness

chriskrycho committed:
Consistency in quote mark style!

chriskrycho committed:
Add first pass on Test Infrastructure

chriskrycho committed:
Finish first section of Migration Docs

chriskrycho committed:
Remove desc. of 'Unresolved questions' section

chriskrycho committed:
Improve structure and finish migration section draft

chriskrycho committed:
Drop needlessly confusing hr in docs sample

chriskrycho committed:
Add example messaging for dropping a TS version

chriskrycho committed:
Reference 0724's edition commitment

chriskrycho committed:
Add an unanswered question

chriskrycho committed:
Add two new sections to Detailed Design

chriskrycho committed:
Some formatting for legibility

chriskrycho committed:
Draft 'Publishing types and release channels' section

chriskrycho committed:
Structural improvements

chriskrycho committed:
Drop redundant 'Rollout' section

chriskrycho committed:
Drop (empty) 'Feature stabilization' section

I thought this was going to be its own thing, but I think it just falls
out naturally from the updates to the RFC process plus all the other
comments on publishing types.

chriskrycho committed:
Draft 'RFC process updates' section

chriskrycho committed:
Add TODOs for subsections in Ember's docs section

chriskrycho committed:
Add lint configuration to CLI Integration section

chriskrycho committed:
Add a couple drawbacks

chriskrycho committed:
More details on changes to CLI/linting

chriskrycho committed:
Draft of 'API documentation' section

chriskrycho committed:
Finish first draft!

muziejus commented:
Like I said in Discord, my imagination regarding the documentation is:

* In the [Orientation](https://guides.emberjs.com/release/tutorial/part-1/orientation/) page there's a mention for something like "use -ts if you want to use TypeScript instead of JavaScript." Similar hints would be scattered in the documentation, like in the configuration section, etc.
* Throughout, code snippets have TS versions and JS versions that are togglable. Similarly, one can set JS or TS for the site as a whole
* There's a special section on TypeScript under "Application Development," which would describe the types ember uses, how to test for types, how to migrate, etc., though I imagine this section would be shortish?

These three switches, it seems, would make TS seem pretty first-class in the documentation. It's always there, but it's not in your face. I think the a11y model might be good to follow. Make it accessible by default (though TS isn't default), and reward people who want to do a deeper dive. In this case, make TS an option in the detailed tutorial and then just move on.

Ideally the code snippets would all be written in TS and then the JS version would be compiled from it, like with the proposed blueprints. This means not having to maintain two versions of each code snippet.

chriskrycho commented:
@muziejus I believe all of that is covered in the RFC. :)

muziejus commented:
Super. Can't wait to contribute to the documentation.

chriskrycho committed:
Fix formatting for details on `noUncheckedIndexAccess`

chriskrycho committed:
Correct the strictness thing

chriskrycho committed:
Some typos and errata

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+- an edition support policy which tells us what our types *must* support (as distinct from what they *may* support)
+- a plan for migrating existing Ember TypeScript users from DefinitelyTyped to using Ember’s own core types
+- a plan for CLI integration (e.g. `ember new --typescript`)
+- test infrastructure for Ember’s type, both to prevent regressions and to catch breaking changes from TypeScript early so they can be mitigated
```
```suggestion
- test infrastructure for Ember’s types, both to prevent regressions and to catch breaking changes from TypeScript early so they can be mitigated
```

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+The only exceptions to the “no new red squiggles” rule are:
+
+- *Bug fixes to the types*: where we previously allowed code we should not have and which could produce runtime errors, the introduction of those red squiggles is making your code safer.
+- *Rare cases involving inferred types*: very rarely, we may make a change to the types which is safe for your code at runtime, and which allows TypeScript to detect new unused code paths or similar. These will introduce “red squiggles” in your editor, but in a way which will simply allow you to clean up code you didn’t need.
```
To head off any FUD here, would it be worthwhile to link to an example of this so folks can see an example of what "code you didn't need" looks like? (I think the semver RFC had one or more of those?)

chriskrycho replied:
Good callout. I'll add some. And yeah: Ryan Cavanaugh found multiple examples of this. :joy:

````suggestion
- *Rare cases involving inferred types*: very rarely, we may make a change to the types which is safe for your code at runtime, and which allows TypeScript to detect new unused code paths or similar. These will introduce “red squiggles” in your editor, but in a way which will simply allow you to clean up code you didn’t need.

    <details><summary>example of this kind of change</summary>

    Assume a library provides a function, `idFor`, which gives back the `id` for a given item. Maybe it finds an `id` field on the object and hands it back if it exists, or generates a UUID otherwise, or similar—the details are unimportant, as we only care about the type signature.
    
    In v1.0 of the library, it specifies that it always returns a `string` or ` a `number`:

    ```ts
    declare function idFor(obj: unknown): string | number;
    ```

    In our own code, we need IDs to always be `string`s, so we do something like this:

    ```ts
    const id = idFor({ cool: 'story' });
    const usableId = typeof id === 'number' ? id.toString() : id;
    ```

    Then let’s say that the authors of `idFor` recognize that returning `string | number` is kind of annoying, and they update their API to always return `string`:

    ```ts
    function idFor(obj: unknown): string;
    ```    
    
    This is perfectly safe from a runtime perspective. Any code which worked before will work now. However, TypeScript is smart enough to see that the `typeof` check is dead code:

    ```ts
    const id = idFor({ cool: 'story' });
    const usableId = typeof id === 'number' ? id.toString() : id; // ❌
    ```

    TypeScript will now report `"Property 'toString' does not exist on type 'never'."` on the `id.toString()` call, because it determines that `id` can never be a `number`, which makes that entire branch dead code. ([playground][id-play]) The "cleanup" work here is that you can simply *get rid of* that code, though!

    For further examples and a discussion of why we cannot work around this (e.g. with lint rules), see [**Appendix C**](https://github.com/chriskrycho/ember-rfcs/blob/semver-for-ts/text/0730-semver-for-ts.md#appendix-c-on-variance-in-typescript) in the Semantic Versioning for TypeScript Types RFC.

    </details>

[id-play]: https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEBbA9sArhBA1AjPA3gFDwnwgAeADsjAC7wBmaAdmLQJbLPzvABiNABTIARgCsAXPBYBrZsgDuzAJRSAzrRjtmAc3gAfeMzSIRIGAG5S8QgF9ChUJFgIU6TPCwAmAsVIVqOkYWNk5uXgEYYXEpWXklVXgNLV0LOwcmVg4ueAA3HEFlX2swLg0eYHgAXk8cADoIoXx4UuQIKQByDRoATw74W2U0krL6NDUoEUwASUqa2h7KEGQGCuqqmo7jU3N+gH4KutpkAGVNbR1C+CleNPtCTNCc3K8rohHmct5qzy8G-iaLWQbU63RgfQGQz8JFKnzGEymIFmPwWSxWaw2m22Zhg+0OxzOKUuRRuwDuhCAA

````

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+#### Strictness
+
+In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultExports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)
```
```suggestion
In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultImports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)
```

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+  [key: string]: string;
+}
+
+function describe(anything: anything) {
```
```suggestion
function describe(anything: Anything) {
```

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+#### Primary packages
+
+The primary packages in the ecosystem, `ember-source`, `ember-data`, and `ember-cli`, will adopt the [rolling window](https://github.com/chriskrycho/ember-rfcs/blob/semver-for-ts/text/0730-semver-for-ts.md#rolling-support-windows) policy:
```
In a future where these packages are broken out into their constituents (which is already partly true with the `@ember-data/*` packages that ship today), is there an easy way as a user to know what's considered a primary package and what isn't?

chriskrycho replied:
Hmmmmmmmmm. This is a very interesting question. I don’t have any *great* thoughts on it, to be honest. I think it’s probably a thing we should tackle if/as that happens? But I could also be persuaded we need to address it here. 🤔 Going to add it to **Unresolved questions**.

chriskrycho replied:
On reflection as I started writing it up, I instead integrated it into **Alternatives: Documentation**. Thank you for calling it out!

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+This does introduce two new small considerations:
+
+- New APIs will now need to specify import paths for type-only imports. For example, today there is no public import for `Transition` or `RouteInfo`, because they are not intended for end users to construct. However, types for them *should* be publicly exported because end users need to be able to name those types them in their own code. If we were to write [RFC #0095: Router Service](https://emb ...
```
❤️

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+### Publishing types and release channels
+
+When Ember official packages begin publishing types, they will need to account for TypeScript's default of using Node module resolution to lookup. In the future, we will be able to publish types in a much more 
```
Missing the end of this sentence?

chriskrycho replied:
Missing most of a paragraph, in fact. :joy: Updated!

dfreeman commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+The outline of that migration path, regardless of the specific package, is:
+
+1. Upgrade to a version of the package (e.g. `ember-source`) which officially publishes types.
+2. Remove the corresponding `@types` package(s) from.
```
```suggestion
2. Remove the corresponding `@types` package(s) from your dependencies.
```

dfreeman reviewed (commented):
Thank you _so_ so much for all your hard work on this!

chriskrycho committed:
Incorporate @dfreeman's errata corrections

Co-authored-by: Dan Freeman <dfreeman@salsify.com>

chriskrycho commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+Beyond this, we *may* define a set of *additional* “lint”-type strictness flags over time taking advantage of TypeScript and/or ESLint, an `ember-ultra-strict` mode, and provide tooling to expose that.
+
+A consequence of adopting `strict: true` as a behavior is that we will be progressively catching *more* type safety issues over time. Since these changes are not gated on TypeScript minor versions, they make cause “red squiggles” during any TypeScript upgrade. However, we still believe this is an appropriate default, for the following reasons:
```
```suggestion
A consequence of adopting `strict: true` as a behavior is that we will be progressively catching *more* type safety issues over time. Since these changes are not gated on TypeScript major versions (that is, they may appear in *any* TypeScript release), they make cause “red squiggles” during any TypeScript upgrade. However, we still believe this is an appropriate default, for the following reasons:
```

chriskrycho commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+#### Strictness
+
+In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], Ember will set `strict: true`, `noUncheckedIndexedAccess: true`. Additionally, we will set `esModuleInterop: false` and `allowSyntheticDefaultImports: false`. These will also be the defaults for Ember projects generated using a `--typescript` flag. (See discussion below under [**CLI Integration**](#cli-integration).)
```
```suggestion
In line with the [Semantic Versioning for TypeScript Types RFC][rfc-0730], both Ember projects and the blueprints generated by Ember using the new `--typescript` flag will set the following compiler options:

- `strict: true`
- `noUncheckedIndexedAccess: true`
- `esModuleInterop: false`
- `allowSyntheticDefaultImports: false`

(See discussion below under [**CLI Integration**](#cli-integration) for further details on the new `--typescript` flag`.)
```

chriskrycho commented on text/0000-ts-adoption-plan.md:
```diff
@@ -0,0 +1,780 @@
+
+> Code which type checks at any given point in a major release of Ember packages will continue to type check throughout that Ember major release for consumers who use supported TypeScript versions.
+
+This is **the “no new red squiggles rule”:** if you use supported TypeScript versions, you will not see new “red squiggles” for type errors in your editor in an Ember major version.
```
```suggestion
This is **the “no new red squiggles rule”:** if you use supported TypeScript versions, you will not see new “red squiggles” for type errors in your editor for Ember minor version upgrades.
```

(I don't love this but haven’t come up with something I *do* love. Input welcome.)

chriskrycho committed:
Fix typos and infelicitous wordings; add an example

chriskrycho committed:
Fix a typo

chriskrycho committed:
Show type error in example better

chriskrycho committed:
Sentence semantics yo

chriskrycho committed:
Clarifications/improvements to alternatives

- Courtesy of @wagenet's work (and in line with the recommended
  migration path for Ember Classic types *in this RFC*!) the bit about
  `Controller` ambient types became defunct, and we should in fact be
  able to rely on Ember’s own types as the full source of truth.
- Fix an incomplete sentence/paragraph caught by @dfreeman (on the
  SemVer option of *not* using `strict: true` but using specific
  flags).

chriskrycho committed:
Fix incomplete thought on 'Publishing types'

chriskrycho committed:
Discuss `@internal` and handling for Ember internals

chriskrycho committed:
Add note on documenting primary/secondary packages

chriskrycho committed:
Improve discussion in 'Ember internals'

chriskrycho committed:
Update publish date to today

chriskrycho committed:
Improve summary

chriskrycho commented:
Opened as emberjs/rfcs#800!
